backend/src/test/java/heartbeat/service/report/ReportServiceTest.java

```diff
@@ -81,71 +84,210 @@ class GenerateReportByType {
 		GenerateReportRequest request = GenerateReportRequest.builder()
 			.csvTimeStamp(timeStamp)
 			.metrics(new ArrayList<>())
+			.metricTypes(List.of("board"))
 			.build();
 
 		@Test
-		void shouldCallGenerateBoardReportWhenMetricTypeIsBoard() throws InterruptedException {
-			MetricsDataCompleted expected = MetricsDataCompleted.builder().boardMetricsCompleted(false).build();
+		void shouldCallGenerateBoardReportAndInitializeMetricDataWhenMetricTypesListOnlyHasBoardElement() {
+			MetricsDataCompleted expected = MetricsDataCompleted.builder()
+				.boardMetricsCompleted(false)
+				.allMetricsCompleted(false)
+				.build();
 			doAnswer(invocation -> null).when(asyncMetricsDataHandler).putMetricsDataCompleted(any(), any());
 			doAnswer(invocation -> null).when(generateReporterService).generateBoardReport(request);
-
-			reportService.generateReportByType(request, MetricType.BOARD);
-			Thread.sleep(100);
-
-			verify(asyncMetricsDataHandler).putMetricsDataCompleted(request.getBoardReportId(), expected);
-			verify(generateReporterService).generateBoardReport(request);
-			verify(generateReporterService, never()).generateDoraReport(request);
+			when(generateReporterService.getComposedReportResponse(any()))
+				.thenReturn(ReportResponse.builder().reportMetricsError(ReportMetricsError.builder().build()).build());
+
+			reportService.generateReport(request);
+			Awaitility.await().atMost(5, TimeUnit.SECONDS).untilAsserted(() -> {
+				verify(asyncMetricsDataHandler)
+					.putMetricsDataCompleted(IdUtil.getDataCompletedPrefix(request.getCsvTimeStamp()), expected);
+				verify(generateReporterService).generateBoardReport(request);
+				verify(generateReporterService, never()).generateDoraReport(request);
+				verify(generateReporterService).getComposedReportResponse(request.getCsvTimeStamp());
+				verify(asyncMetricsDataHandler)
+					.updateAllMetricsCompletedInHandler(IdUtil.getDataCompletedPrefix(request.getCsvTimeStamp()));
+			});
 		}
 
 		@Test
-		void shouldCallGenerateDoraReportWhenMetricTypeIsDora() throws InterruptedException {
-			MetricsDataCompleted expected = MetricsDataCompleted.builder().doraMetricsCompleted(false).build();
+		void shouldCallGenerateDoraReportWhenMetricTypesListOnlyHasDoraElement() {
```

Comment:
- description
- JUnit tests should include assert() or fail()

Reply:
The test method is created into two threads, with no return value, only to verify the input and number of calls to the method within the thread

Reply:
done,  have been updated